### DISCUSSION TITLE
Force triangulation along the z axis

### DISCUSSION THREAD

**Discussion by elifoo** [pyvista/pyvista#1491] [Q&A] [answered]
Hi,

I'm trying to use the delaunay_3d () function to get a 3D triangulation of points (fig.1-2, file 'point_xyz.npy.zip') that are arranged by slice along z (same distance along z).

<img width="565" alt="fig 1" src="https://user-images.githubusercontent.com/78592327/124495589-95b01d00-ddb8-11eb-942b-5f0597f910f2.png">

<img width="625" alt="fig 2" src="https://user-images.githubusercontent.com/78592327/124495617-9ea0ee80-ddb8-11eb-8076-5b13b40f9644.png">

[point_xyz.npy.zip](https://github.com/pyvista/pyvista/files/6765321/point_xyz.npy.zip)

Unfortunately, the triangulation does not follow the distance of the slices and the mesh thus obtained has irregular triangles (fig. 3).

<img width="413" alt="fig  3" src="https://user-images.githubusercontent.com/78592327/124495678-b2e4eb80-ddb8-11eb-9204-ee630229cdf4.png">

To achieve this, I have already tried to use the Triangulation function of the matplotlib.tri library (fig. 4),

<img width="565" alt="fig  4" src="https://user-images.githubusercontent.com/78592327/124495703-bc6e5380-ddb8-11eb-9996-59ad04818e5f.png">

 but I have some problems both in the top and bottom closure (fig. 5) and in the total closure of the surface (see fig. 6).

<img width="498" alt="fig  5" src="https://user-images.githubusercontent.com/78592327/124495747-c6905200-ddb8-11eb-9b17-587f36b99d49.png">

<img width="499" alt="fig  6" src="https://user-images.githubusercontent.com/78592327/124495770-cf812380-ddb8-11eb-9e65-6a7a6ef1e1d2.png">

Any idea how I can fix it?

Thanks a lot

**Comment by akaszynski** [comment, marked as answer]
I'd feed in the result from `matplotlib.tri` into `pyvista` and then run [fill_holes](https://docs.pyvista.org/core/filters.html#pyvista.PolyDataFilters.fill_holes) to clean up the mesh.

**Reply by elifoo** [reply]
Hello,
I am trying to extract the mesh from matplotlib to pass it to pyvista, but I have a numpy array of shape (n,3,3) (vertices information). How do I get a polydata? I tried with pyvista.vtkmatrix_from_array (my numpy array), but it only accepts shape of  (3,3) or (4,4).

But no other solution to improve the pyvista mesh? Is it possible to create a mesh with pyvista with polar coordinates?

Thanks!

**Reply by akaszynski** [reply]
According to https://matplotlib.org/stable/api/tri_api.html, you're provided with an array of faces, what you need to do is construct a VTK compatible array via something like:

```py
import matplotlib
import pyvista as pv
import numpy as np

np_points = np.load('point_xyz.npy')

from matplotlib.tri import Triangulation
tri = Triangulation(np_points[:, 0], np_points[:, 1],
                    triangles=None, mask=None)
triangles = tri.triangles

# from scipy.spatial import Delaunay
# tri = Delaunay(np_points)
nfaces = triangles.shape[0]

faces = np.empty((nfaces, 4), dtype=int)
faces[:, 0] = 3
faces[:, 1:] = triangles[:, :3]

mesh = pv.PolyData(np_points, faces)
mesh.plot(show_edges=True)
```

The mesh generated here isn't correct; I don't know what settings you're using to generate your triangulation within `matplotlib`, but this should help you create your mesh within `pyvista`.
